crates/sui-rpc-benchmark/src/json_rpc/runner.rs

```diff
@@ -54,51 +76,179 @@ impl JsonRpcStats {
-            let is_error = !matches!(res, Ok(Ok(ref resp)) if resp.status().is_success());
-
-            let mut stats = task_stats
-                .lock()
-                .map_err(|e| anyhow::anyhow!("Failed to acquire stats lock: {}", e))?;
-            stats.record_request(&request_line.method, elapsed_ms, is_error);
-            Ok::<(), anyhow::Error>(())
-        }
-    }));
+    let tasks = futures::stream::iter(requests.into_iter()).try_for_each_spawned(
+        concurrency,
+        |mut request_line| {
+            let task_stats = stats.clone();
+            let client = client.clone();
+            let endpoint = endpoint.clone();
+            let pagination_state = pagination_state.clone();
+
+            // adapt pagination cursor to new cursor format if needed
+            async move {
+                let params = request_line
+                    .body_json
+                    .get("params")
+                    .and_then(|v| v.as_array())
+                    .map(|a| a.to_vec())
+                    .with_context(|| {
+                        format!(
+                            "params not found or not an array for method: {}",
+                            request_line.method
+                        )
+                    })?;
+
+                if let Some(cursor_idx) =
+                    PaginationCursorState::get_method_cursor_index(&request_line.method)
+                {
+                    let method_key =
+                        PaginationCursorState::get_method_key(&request_line.method, &params)?;
+                    PaginationCursorState::update_params_cursor(
+                        &mut request_line.body_json,
+                        cursor_idx,
+                        pagination_state.get(&method_key).as_ref(),
+                        &request_line.method,
+                    )?;
+                }
+
+                let now = Instant::now();
+                let res = timeout(
+                    Duration::from_secs(10),
+                    client.post(&endpoint).json(&request_line.body_json).send(),
+                )
+                .await;
+                let elapsed_ms = now.elapsed().as_millis() as f64;
+
+                // update pagination cursor if the request is successful.
+                let mut is_error = true;
+                if let Ok(Ok(resp)) = res {
+                    if resp.status().is_success() {
+                        #[derive(Deserialize)]
+                        struct Body {
+                            result: Result,
+                        }
+                        #[derive(Deserialize)]
+                        #[serde(rename_all = "camelCase")]
+                        struct Result {
+                            has_next_page: bool,
+                            next_cursor: Option<Value>,
+                        }
+
+                        if let Ok(Body { result }) = resp.json::<Body>().await {
```

Comment:
Do you need the type annotation if you are destructuring the type?

```suggestion
                        if let Ok(Body { result }) = resp.json().await {
```